test/shared/src/main/scala/zio/test/TestAspect.scala

```diff
@@ -439,18 +456,67 @@ object TestAspect extends TimeoutVariants {
   def parallelN(n: Int): TestAspectPoly = executionStrategy(ExecutionStrategy.ParallelN(n))
 
   /**
-   * An aspect that retries failed tests according to a schedule.
+   * An aspect that restores a given [[Restorable]]'s state to its starting state after the test is run.
+   * Note that this is only useful when repeating tests.
    */
-  def retry[R0, E0, S0](
-    schedule: Schedule[R0, TestFailure[E0], S0]
-  ): TestAspect[Nothing, R0, Nothing, E0, Nothing, Any] =
-    new TestAspect.PerTest[Nothing, R0, Nothing, E0, Nothing, Any] {
-      def perTest[R >: Nothing <: R0, E >: Nothing <: E0, S >: Nothing <: Any](
+  def restore[R0 <: Restorable](r: R0): TestAspect[Nothing, R0, Nothing, Any, Nothing, Any] =
```

Comment:
@adamgfraser I did initially look at `around` but for the reasons you mentioned moved to `r.flatmap(test <* _)`. I do like the solution for generalising `around` though - it'll be much more useful to everyone like that.

RE the service being `Restorable` and not the e.g. `TestClock`,  `save` has the same signature on the service and the helper methods - I thought I should be able to do something like this:
```scala
trait TestClock extends Clock with Scheduler with Restorable {
   ???
}

def restore[R0 <: Restorable](service:R0): TestAspect[Nothing, R0, Nothing, Any, Nothing, Any] = ???

val restoreTestClock: TestAspect[Nothing, TestClock, Nothing, Any, Nothing, Any] = restore(TestClock)
```
but the compiler said no:
```
[error] C:\Users\rober\IdeaProjects\zio\test\shared\src\main\scala\zio\test\TestAspect.scala:447:93: type mismatch;
[error]  found   : zio.test.TestAspect[Nothing,R0,Nothing,Any,Nothing,Any]
[error]  required: zio.test.TestAspect[Nothing,zio.test.environment.TestClock,Nothing,Any,Nothing,Any]
[error]   val restoreTestClock: TestAspect[Nothing, TestClock, Nothing, Any, Nothing, Any] = restore(TestClock)
```
is there a fundamental reason this can't work that I'm not understanding?

your approach of `service: R0 => Restorable` does work but doesn't feel right to me when we have helper methods on e.g. `TestClock` which aren't even used

Reply:
The issue is with the distinction between the companion object and the trait, and the service.  In your snippet `restore` is looking for a value that extends `Restorable` but neither the `TestClock` trait or the companion object `TestClock` implement `Restorable`.

I don't think you can implement `Restorable` at the level of the companion object because you need to get the environment to access the `TestClock` you want to restore so you can't satisfy the type signature of `Restorable`. I think you could potentially make `TestClock` the trait extend `Restorable` (the implementation would just call the `restore` method on the service). But I don't think it would be a good idea because when you mixed together different traits (e.g. `TestClock with TestConsole` you would create issues of which implementation of `Restorable` to use.

Reply:
> But I don't think it would be a good idea because when you mixed together different traits (e.g. TestClock with TestConsole you would create issues of which implementation of Restorable to use.

I didn't think of this - I'll  leave it how it is